### DISCUSSION TITLE
Wemos D1 mini support

### DISCUSSION THREAD

**Discussion by Tardymo** [raomin/ESPAltherma#218] [General]
A couple of years ago I bought Wemos D1 mini and D1 mini Pro modules. Since they are ESP8266 based, I was hoping to be able to use it for this project. In platformio.ini I have added this configuration:
```
[env:d1_mini_pro]
platform = espressif8266
framework = arduino
board = d1_mini_pro
build_flags = -D_GNU_SOURCE
lib_deps =
	PubSubClient
```
I am able to build and upload successfully, but the device never connects to my WiFi. I have made the necessary changes in setup.h, so I am not sure what to do next. I am a C developer myself, but have no experience with ESP chips, platformio, nodemcu e.t.c, so I am hoping for some help to get me started here. It appears D1 mini can not be debugged in PlatformIO, since there is a error message:  undefinedError: Currently, PlatformIO does not support debugging for `WeMos D1 R2 and mini`.

**Comment by Tardymo** [comment]
I had to also add monitor_speed = 115200 to see terminal output. There I see, that the device is in reboot loop, with this output:

```
 ets Jan  8 2013,rst cause:2, boot mode:(3,6)

load 0x4010f000, len 3584, room 16
tail 0
chksum 0xb0
csum 0xb0
v2843a5ac
~ld
```
I am afraid I will not be able to solve this myself. Has this project been tested on ESP8266 recently? Maybe some libraries changed causing this reboot loop?

**Comment by Tardymo** [comment]
I decided to have a go at finding where the problem is. So I commented out everything in `void setup()` and `void loop()`, and it still the same problem. So I narrowed down the problem to including "setup.h". And the problem is with PROGMEM macro which is used for labelDefs. I removed that and now everything works. I got a successful connection to WiFi and mqtt broker. So now I have to connect the board to my Altherma unit to query some values. I will probably use a voltage divider for Rx pin as I don't think driving a pin with 5V is a good idea.

So what is a deal with PROGMEM? If I understand correctly, it is used to put labelDefs in program memory instad of RAM. But why isn't it working on esp8266?

**Reply by raomin** [reply]
Hi @Tardymo, welcome! 

I haven't been using esp8266 for ESPAltherma, that's why it might not be the most reliable flavour of the code, sorry for that. Now it looks from what I've seen that this implementation could need a _flashstringhelper_ function to properly get back the value from flash.
Now I also read that on non AVR using `const` should be enough to store big data in flash.
Anyway, you can definitely remove all PROGMEM statements from the code, it should just work fine.

I'll then have a more thorough look at it to see if it has any advantages at all or if I should simply remove it for the esp8266 or even for all...

As for the voltage deviser, I'm not aware of any of us using one. ESPs are pretty 5v tolerant.

Cheers

**Comment by wikipeter-de** [comment]
Hi @raomin, 

I had the same problem with the boot loop with my ESP8266. When I remove the PROGMEM from "ALTHERMA(HYBRID).h" I only get this by MQTT: 
{"WifiRSSI":"-73dBm","FreeMem":"32040"}

Although I uncommented some values:

```
LabelDef labelDefs[] = {
...
{0x62,8,302,1,-1,"Circulation pump operation"},
//{0x62,8,301,1,-1,"Alarm output"},
{0x62,8,300,1,-1,"Space C/H Op. or BPH"},
{0x62,9,105,2,-1,"Flow sensor (l/min)"},
{0x62,11,105,1,2,"Water pressure"},
{0x62,12,152,1,-1,"Water pump signal (0:max-100:stop)"},
...
//{0x00,0,996,0,-1,"Override All Clear"},
};
```
... But definitely a cool project! And so I stay tuned and try to get it working. Even if I order a ESP32. ;-)

**Reply by RobertSzabo** [reply]
Hi, just run into the same problem with Wemos D1 mini. At least, without PROGMEM I could get rid of the boot loops. I will try to connect to Daikin heat pump later. BTW, I had version mismatch between the platform and the serial libraries too when compiling. 
Thanks for the great work!

**Comment by swalsh11** [comment]
Did anyone manage to get a Wemos ESP8266 D1 to work?

**Reply by Tardymo** [reply]
Yes, I am running it currently and the solution is given above - remove PROGMEM from label definition header file.

**Reply by wikipeter-de** [reply]
> Did anyone manage to get a Wemos ESP8266 D1 to work?

No, I changed to ESP32. That was easier to do.

**Comment by swalsh11** [comment]
Thanks @Tardymo

**Comment by TRIROG** [comment]
Works fine in my setup - i did the compiling and uploading through platrofmIO as per guide.
